kb/testing/testing-constant-flow.md

```diff
@@ -0,0 +1,99 @@
+
+### Valgrind's memcheck
+
+It turns out patching valgrind and using a helper library is not necessary, as everything we need is already provided by memcheck: after including `valgrind/memcheck.h` we can use the macros `VALGRIND_MAKE_MEM_UNDEFINED(addr, len)` and `VALGRIND_MAKE_MEM_DEFINED(addr, len)` to mark data as secret (undefined) or no longer secret (defined).
+
+These macros have been available under that name since [3.2.0 (7 June 2006)](https://www.valgrind.org/docs/manual/dist.news.old.html).
+
+References:
+- https://valgrind.org/docs/manual/mc-manual.html
+- https://github.com/mozilla-b2g/valgrind/blob/master/memcheck/memcheck.h
+
+Pros:
+- valgrind/memcheck is already used in our CI
+- large number of [supported architectures/platforms](https://valgrind.org/info/platforms.html)
+- works with any compiler
+- works with inline asm
+- already used by other projects such as [in BoringSSL](https://boringssl.googlesource.com/boringssl/+/a6a049a6fb51a052347611d41583a0622bc89d60)
+
+Cons:
+- just like normal use of valgrind's memcheck, slows down execution by about 20-30x.
+- preliminary testing (by Manuel on a personal project) suggests there may be issues with relative ordering of calls to the macros and use of the values in some uses cases (for example, doing a simple sanity check before marking the value as secret flags the sanity check as accessing undefined memory).
+
+Note: since 2.24.0 we're using this: see `MBEDTLS_TEST_CONSTANT_FLOW_VALGRIND` in `config.h`; components `test_valgrind_constant_flow` and `test_valgrind_constant_flow_psa` in `all.sh`; and `tests/suites/test_suite_constant_time`.
+
+### MemSan
+
+Works in a very similar way to valgrind's memcheck: compiling with `clang -fsanitize=memory`, after including `sanitizer/msan_interface.h`, the two following functions (and a couple more) can be used to mark memory areas as secret/undefined or public/defined.
+```
+  /* Make memory region fully initialized (without changing its contents). */
+  void __msan_unpoison(const volatile void *a, size_t size);
+
+  /* Tell MSan about newly allocated memory (ex.: custom allocator).
+     Memory will be marked uninitialized, with origin at the call site. */
+  void __msan_allocated_memory(const volatile void* data, size_t size);
+```
+
+References:
+- https://clang.llvm.org/docs/MemorySanitizer.html
+- https://github.com/google/sanitizers/wiki/MemorySanitizer
+- https://github.com/llvm-mirror/compiler-rt/blob/master/include/sanitizer/msan_interface.h
+
+Pros:
+- clang/memsan is already used in our CI
+- much faster than valgrind/memcheck (only 2-3x slowdown, so about 10x faster than using valgrind)
+
+Cons:
+- only works with clang on Linux, FreeBSD, NetBSD
+- being compile-time instrumentation, limited with respect to inline assembly
+- for the same reason, misses branches in toolchain-provided functions (for example long multiplication, see [this paper](https://core.ac.uk/download/pdf/188281541.pdf)) which valgrind would see
```

Comment:
The link to the paper is broken here and I am afraid I am not familiar with which paper is supposed to be linked here.

Reply:
Searching core.ac.uk for that number yields https://core.ac.uk/search?q=188281541 which looks like it matches the description here. @mpg does that look like the right link to you?

Reply:
I am not Manuel, but this certainly seems to match. We should probably cite papers properly, rather than just providing links that may go stale (or at least ask the Wayback machine to archive them!)

Reply:
I've linked it properly.